addons/stock_delivery/models/stock_move.py

```diff
@@ -41,7 +41,16 @@ def _cal_move_weight(self):
     def _get_new_picking_values(self):
         vals = super(StockMove, self)._get_new_picking_values()
         carrier_id = self.group_id.sale_id.carrier_id.id
-        vals['carrier_id'] = any(rule.propagate_carrier for rule in self.rule_id) and carrier_id
+        carrier_tracking_ref = False
+        if self.move_orig_ids and self.move_orig_ids.picking_id:
+            # check if previous picking have carrier_id take carrier from that
+            # earlier we were taking carrier from sale but since carrier can be changed  or updated in next steps so now we take carrier from prev picking
+            carrier_id = self.move_orig_ids.picking_id.carrier_id.id
+            carrier_tracking_ref = self.move_orig_ids.picking_id.carrier_tracking_ref
+        # propagating carrier and tracking ref only if carrier propagation rule allow
+        if any(rule.propagate_carrier for rule in self.rule_id):
+            vals['carrier_tracking_ref'] = carrier_tracking_ref
+            vals['carrier_id'] = carrier_id
```

Comment:
```suggestion
        carrier_tracking_ref = False
        if self.move_orig_ids.picking_id.carrier_id:
            # check if previous picking have carrier_id take carrier from that
            # earlier we were taking carrier from sale but since carrier can be changed  or updated in next steps so now we take carrier from prev picking
            carrier_id = self.move_orig_ids.picking_id.carrier_id.id
            carrier_tracking_ref = self.move_orig_ids.picking_id.carrier_tracking_ref
        # propagating carrier and tracking ref only if carrier propagation rule allow
        if any(rule.propagate_carrier for rule in self.rule_id):
            vals['carrier_tracking_ref'] = carrier_tracking_ref
            vals['carrier_id'] = carrier_id
```